Add queue-level JobExclusiveAllocation cluster configuration parameter

EddyMM opened the pull request:
### Description of changes
* This parameter will be used to optionally configure Slurm Partitions to run jobs on exclusively allocated nodes (OverSubscribe=NO)

```
Scheduling:
  Scheduler: ...
  SlurmQueues:
  - Name: ...
    JobExclusiveAllocation: true | false (default: false)
    ...
```

### Tests
* Unit tests to confirm the parameter is parsed and included in the cluster config model (including the default).
* Manually tested against https://github.com/aws/aws-parallelcluster-cookbook/pull/2294


```
» pcluster create-cluster -c jls-config.yaml --debug -n Cluster-JLS
{
  "configurationValidationErrors": [
    {
      "level": "ERROR",
      "type": "CustomSlurmSettingsValidator",
      "message": "Using the following custom Slurm settings at Queue level is not allowed: OverSubscribe"
    }
  ],
  "message": "Invalid cluster configuration."
}
```

### References
* https://github.com/aws/aws-parallelcluster-cookbook/pull/2294

### Checklist
- [x] Make sure you are pointing to **the right branch** and add a label in the PR title (i.e. **2.x** vs **3.x**)
- [x] Check all commits' messages are clear, describing what and why vs how.
- [x] Make sure **to have added unit tests or integration tests** to cover the new/modified code.
- [x] Check if documentation is impacted by this change.

Please review the [guidelines for contributing](../CONTRIBUTING.md) and [Pull Request Instructions](https://github.com/aws/aws-parallelcluster/wiki/Git-Pull-Request-Instructions).

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

lukeseawalker commented on cli/src/pcluster/schemas/cluster_schema.py:
```diff
@@ -1559,6 +1559,7 @@ class SlurmQueueSchema(_CommonQueueSchema):
     tags = fields.Nested(
         QueueTagSchema, many=True, metadata={"update_policy": UpdatePolicy.QUEUE_UPDATE_STRATEGY, "update_key": "Key"}
     )
+    job_exclusive_instance_allocation = fields.Bool(metadata={"update_policy": UpdatePolicy.QUEUE_UPDATE_STRATEGY})
```
what about a more relaxed update policy?

EddyMM replied:
+1.
`UpdatePolicy.SUPPORTED` should work.

lukeseawalker commented on cli/tests/pcluster/schemas/test_cluster_schema.py:
```diff
@@ -1435,3 +1435,53 @@ def test_queue_tag_schema(mocker, config_dict, failure_message):
+        ),
+    ],
+)
+def test_job_exclusive_instance_allocation_deserialization(
```
what about removing "_deserialization" suffix?

lukeseawalker commented on cli/tests/pcluster/schemas/test_cluster_schema.py:
```diff
@@ -1435,3 +1435,53 @@ def test_queue_tag_schema(mocker, config_dict, failure_message):
+                "Name": "Queue",
+                "Networking": {"SubnetIds": ["subnet-12345678"]},
+                "ComputeResources": [{"Name": "compute_resource", "InstanceType": "t2.micro"}],
+                "JobExclusiveInstanceAllocation": True,
```
What about  `JobExclusiveInstanceAllocation` = "true", "false", "", 0, 1 and None?

EddyMM replied:
Added.
It's worth noting that Marshamallow [fields.Bool](https://marshmallow.readthedocs.io/en/stable/marshmallow.fields.html#marshmallow.fields.Boolean) uses "truthiness" in a very general sense i.e. it evaluates the truthiness of a value, instead of only True/False. We use the same field type in other params as well e.g. **Efa/Enabled** can work with `1`.

EddyMM committed:
Add queue-level JobExclusiveAllocation cluster configuration parameter

This parameter will be used to optionally configure Slurm Partitions to run jobs on exclusively allocated node (OverSubscribe=NO)

Signed-off-by: Eddy Mwiti <eddmwiti@amazon.com>
